src/utils/fhir.js

```diff
@@ -60,4 +134,92 @@ function refreshKnowledgeArtifacts(db) {
   });
 }
 
-module.exports = { generateOperationOutcome, refreshKnowledgeArtifacts };
+/**
+ * Take a Knowledge Artifact Bundle and generate Subscription resources for the named events
+ *
+ * @param {Bundle} specBundle - KA Bundle with PlanDefinition
+ * @param {string} url - the notification endpoint url
+ * @param {string} token - the authorization token for the Subscription notification
+ * @returns list of Subscription resources
+ */
+function subscriptionsFromBundle(specBundle, url, token) {
+  const planDefs = specBundle.entry.filter(e => e.resource.resourceType === 'PlanDefinition');
+  if (!planDefs.length) {
+    throw new Error('Specification Bundle does not contain a PlanDefinition');
+  }
+
+  return planDefs.map(planDef => subscriptionsFromPlanDef(planDef.resource, url, token)).flat();
+}
+
+/**
+ * Take a single PlanDefinition and generate a Subscription resource for the named events
+ *
+ * @param {PlanDefinition} planDef - KA PlanDefinition
+ * @param {string} url - the notification endpoint url
+ * @param {string} token - the authorization token for the Subscription notification
+ * @returns list of Subscription resources (one for each trigger on the first action).
+ */
+function subscriptionsFromPlanDef(planDef, url, token) {
+  const action = planDef.action[0];
+
+  if (action.trigger?.length === 0) return [];
+
+  return action.trigger.map(trigger => {
+    const namedEventExt = trigger.extension.find(e => e.url === NAMED_EVENT_EXTENSION);
+    const namedEventCoding = namedEventExt.valueCodeableConcept.coding.find(
+      c => c.system === NAMED_EVENT_CODE_SYSTEM
+    );
+
+    const criteria = namedEventToCriteria(namedEventCoding.code);
+    if (criteria) return generateSubscription(namedEventCoding.code, criteria, url, token);
```

Comment:
If `criteria` is null then this inner function returns undefined, and we'll probably have to check for that in the array downstream. Can this be reworked to ensure the function doesn't pull null/undefined into the list? Or another quick approach is just tack `.filter(x => x)` after the `.map(...)` call to filter out any falsy values.

The other question this raises for me is, should there be some way of indicating that we have a PlanDef that we can't create a subscription for, because we can't map it to criteria here? Maybe for now just write to debug output that it happened?

Reply:
Updated the map to be a reduce and wrote a new test to verify it works